src/Passes/docs/base-profile-transformations.md

```diff
@@ -0,0 +1,233 @@
+
+```
+define internal void @Feasibility__QubitMapping__body() {
+entry:
+  %qs = call %Array* @__quantum__rt__qubit_allocate_array(i64 3)
+  call void @__quantum__rt__array_update_alias_count(%Array* %qs, i32 1)
+  br label %header__1
+
+header__1:                                        ; preds = %exiting__1, %entry
+  %q = phi i64 [ 8, %entry ], [ %4, %exiting__1 ]
+  %0 = icmp sle i64 %q, 10
+  br i1 %0, label %body__1, label %exit__1
+
+body__1:                                          ; preds = %header__1
+  %1 = sub i64 %q, 8
+  %2 = call i8* @__quantum__rt__array_get_element_ptr_1d(%Array* %qs, i64 %1)
+  %3 = bitcast i8* %2 to %Qubit**
+  %qubit = load %Qubit*, %Qubit** %3, align 8
+  call void @__quantum__qis__x__body(%Qubit* %qubit)
+  br label %exiting__1
+
+exiting__1:                                       ; preds = %body__1
+  %4 = add i64 %q, 1
+  br label %header__1
+
+exit__1:                                          ; preds = %header__1
+  call void @__quantum__rt__array_update_alias_count(%Array* %qs, i32 -1)
+  call void @__quantum__rt__qubit_release_array(%Array* %qs)
+  ret void
+}
+```
+
+After applying the our [demo profile transformation](https://github.com/troelsfr/qsharp-compiler/tree/feature/profile-pass/src/Passes/examples/QubitAllocationAnalysis), the QIR is reduced to:
+
+```
+define void @Feasibility__QubitMapping__Interop() local_unnamed_addr #0 {
+entry:
+  call void @__quantum__qis__x__body(%Qubit* null)
+  call void @__quantum__qis__x__body(%Qubit* nonnull inttoptr (i64 1 to %Qubit*))
+  call void @__quantum__qis__x__body(%Qubit* nonnull inttoptr (i64 2 to %Qubit*))
+  ret void
+}
+```
+
+We note that we successfully have eliminated loops, arithmetic operations, dynamic qubit allocation and alias counting - all operations which are not supported by the base profile.
+
+## Goal
+
+We envision the tool to work as a standalone commandline tool which can either validate or generate a QIR in accordance with a given profile. To validate, one would run:
+
+```language
+qat -p profile.yaml --validate unvalidated-qir.ll
+```
+
+In a similar fashion, generation is performed by adding `--generate` to the commandline:
+
+```language
+qat -p profile.yaml --generate qir.ll > qir-profile.ll
+```
+
+Default behaviour of the tool is that it always validates the generated profile. This behaviour can be disabled by
+
+```language
+qat -p profile.yaml --generate --no-validate qir.ll > qir-profile.ll
+```
+
+# Profile Specification
+
+Every profile is specified through a YAML file which defines an object at the top-level. This object must contain the fields `name` and `displayName`:
+
+```yaml
+name: profile-name
+displayName: Profile Name
+# ...
+```
+
+Additionally, top level also contains the fields `version` and `mode`. The version refers to the QIR version which forms the basis for the specification and `mode` explains how the profile is defined.
```

Comment:
What is your thinking regarding versioning; would you expect that the same library handles all supported versions, or would you imaging that we release a library (and updates) per version? Would people download the correct version or would you imaging the tool to pull the right version (if not already on disk) depending on what is needed?

Reply:
The thought is that at some point the spec will need to mature to cover what the industry will need for the next 3 - 5 years. Once that happens we are at version 1.0 and freeze all specs such that they will not have to build tool chains around work-in-progress. We can then look ahead for what is need 5 - 10 years down the road and start on version 2.0 if we believe that is needed. You can then still make minor changes with minor version increases such as 1.1

I believe this will be needed as building the full hardware setup is a slow and expensive process and will require some level of stability.